### ISSUE TITLE
A Proposal for Keys

### ISSUE THREAD

**Issue by wycats** [toml-lang/toml#220]
This is related to #65, #67, #185, #90, #180, #62, #126, #83, and probably others.

# Motivation

The TL;DR is that keys are currently slightly ambiguous, but also don't support a number of commonly desired characters.

A couple of examples:

```toml
[ips.127.0.0.1]
[directories.Space Separated]
```

In Cargo, this comes up because we use dependency names as keys:

```toml
[dependencies]

hammer.rs = "1.0.0"

# or

[dependencies.hammer.rs] = "1.0.0"
```

# Proposal

* Unadorned keys are one or more [alphabetic](http://www.unicode.org/reports/tr44/#Alphabetic), [numeric value](http://www.unicode.org/reports/tr44/#Numeric_Value), `-`, `_`, `$`, `!` or `?` characters.
* Quoted keys are valid [Toml Strings](https://github.com/toml-lang/toml#string).

So this would be valid:

```toml
[dependencies]

"hammer.rs" = "1.0.0"

# or

[dependencies."hammer.rs"] = "1.0.0"
```

I know there have been many discussions on this topic before. I have read them and have tried to include their considerations in this proposal.

One note: I think including `-` as a valid unadorned identifier character is extremely important.

**Comment by mk-pmb** [comment]
So with current TOML, can I have unicode characters 0 to 31 a key name? I'm not familiar enough with TOML yet to guess about `\n`, but `\r\t` and terminal control escape sequences will surely add to my experience of editing config files on a terminal with grep/sed/cat. Especially for terminals that support silent/password input mode.

> Name your tables whatever crap you please, just don't use a dot. Dot is reserved. OBEY.

Probably well intentioned, but I'd still favor a whitelist approach. With a blacklist, many webdevs might remember to care about non-breaking space, some who know that JSON is not a JavaScript subset may even remember about U+2028 (line separator) and U+2029 (paragraph separator), and might hope that Unicode Consortium won't add too fancy new whitespace ever.

> Or, to keep it simple: key parts must be comprised of arbitrary sequences of characters belonging to the Unicode Categories Letter (L.), Number (N.) and Mark (M.), as well as _ and - (underscore, hyphen).

I couldn't find a good enough twin for `.` in just letters and digits yet, so maybe it's safe (for now) to whitelist all letters and digits. Depending on font, U+05C5 (hebrew mark lower dot) might work in some terminals and editors.

**Comment by mojombo** [comment]
@mk-pmb Please see #283 for the latest proposal on this matter.

**Comment by mk-pmb** [comment]
thanks!

**Comment by mojombo** [comment]
Resolved by #283.
